File: python/leanfe/compress.py
"""
YOCO (You Only Compress Once) compression strategy for leanfe.

Implements the optimal data compression strategy from Wong et al. (2021):
"You Only Compress Once: Optimal Data Compression for Estimating Linear Models"

This strategy compresses data by grouping on (regressors + fixed effects),
computing sufficient statistics (n, sum_y, sum_y_sq), and running weighted
least squares on the compressed data.

Used automatically when vcov is "iid" or "HC1" (not cluster) and no IV.
"""

import numpy as np
import polars as pl
from typing import List, Optional, Dict, Tuple


def should_use_compress(vcov: str, has_instruments: bool) -> bool:
    """
    Determine if compression strategy should be used.
    
    Compression is used when:
    - vcov is "iid" or "HC1" (cluster requires different handling)
    - No instrumental variables
    
    Parameters
    ----------
    vcov : str
        Variance-covariance type
    has_instruments : bool
        Whether IV/2SLS is being used
        
    Returns
    -------
    bool
        True if compression should be used
    """
    vcov_ok = vcov.lower() in ("iid", "hc1")
    return vcov_ok and not has_instruments


def compress_polars(
    df: pl.DataFrame,
    y_col: str,
    x_cols: List[str],
    fe_cols: List[str],
    weights: Optional[str] = None
) -> Tuple[pl.DataFrame, int]:
    """
    Compress data using GROUP BY on regressors + fixed effects.
    
    Parameters
    ----------
    df : pl.DataFrame
        Input data
    y_col : str
        Dependent variable column
    x_cols : list of str
        Regressor columns
    fe_cols : list of str
        Fixed effect columns
    weights : str, optional
        Weight column name
        
    Returns
    -------
    tuple
        (compressed_df, n_obs_original)
    """
    group_cols = x_cols + fe_cols
    n_obs_original = len(df)
    
    if weights is not None:
        # Weighted compression
        agg_exprs = [
            pl.col(weights).sum().alias("_n"),
            (pl.col(y_col) * pl.col(weights)).sum().alias("_sum_y"),
            (pl.col(y_col).pow(2) * pl.col(weights)).sum().alias("_sum_y_sq"),
        ]
    else:
        agg_exprs = [
            pl.len().alias("_n"),
            pl.col(y_col).sum().alias("_sum_y"),
            pl.col(y_col).pow(2).sum().alias("_sum_y_sq"),
        ]
    
    compressed = df.group_by(group_cols).agg(agg_exprs)
    
    # Add mean_y and sqrt weights for WLS
    compressed = compressed.with_columns([
        (pl.col("_sum_y") / pl.col("_n")).alias("_mean_y"),
        pl.col("_n").sqrt().alias("_wts"),
    ])
    
    return compressed, n_obs_original


def compress_duckdb(
    con,
    y_col: str,
    x_cols: List[str],
    fe_cols: List[str],
    weights: Optional[str] = None
) -> Tuple[any, int]:
    """
    Compress data using SQL GROUP BY.
    
    Parameters
    ----------
    con : duckdb.Connection
        DuckDB connection with 'data' table
    y_col : str
        Dependent variable column
    x_cols : list of str
        Regressor columns
    fe_cols : list of str
        Fixed effect columns
    weights : str, optional
        Weight column name
        
    Returns
    -------
    tuple
        (compressed_df as pandas, n_obs_original)
    """
    group_cols = x_cols + fe_cols
    group_cols_sql = ", ".join(group_cols)
    
    n_obs_original = con.execute("SELECT COUNT(*) FROM data").fetchone()[0]
    
    if weights is not None:
        query = f"""
        SELECT
            {group_cols_sql},
            SUM({weights}) AS _n,
            SUM({y_col} * {weights}) AS _sum_y,
            SUM(POWER({y_col}, 2) * {weights}) AS _sum_y_sq,
            SUM({y_col} * {weights}) / SUM({weights}) AS _mean_y,
            SQRT(SUM({weights})) AS _wts
        FROM data
        GROUP BY {group_cols_sql}
        """
    else:
        query = f"""
        SELECT
            {group_cols_sql},
            COUNT(*) AS _n,
            SUM({y_col}) AS _sum_y,
            SUM(POWER({y_col}, 2)) AS _sum_y_sq,
            SUM({y_col}) / COUNT(*) AS _mean_y,
            SQRT(COUNT(*)) AS _wts
        FROM data
        GROUP BY {group_cols_sql}
        """
    
    compressed_df = con.execute(query).fetchdf()
    return compressed_df, n_obs_original


def build_design_matrix(
    compressed_df,
    x_cols: List[str],
    fe_cols: List[str],
    backend: str = "polars"
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, List[str], int]:
    """
    Build design matrix from compressed data with FE dummies.
    
    Parameters
    ----------
    compressed_df : DataFrame
        Compressed data (polars or pandas)
    x_cols : list of str
        Regressor columns
    fe_cols : list of str
        Fixed effect columns
    backend : str
        "polars" or "duckdb"
        
    Returns
    -------
    tuple
        (X, Y, wts, all_col_names, n_fe_levels)
    """
    if backend == "polars":
        # Convert to pandas for easier manipulation
        pdf = compressed_df.to_pandas()
    else:
        pdf = compressed_df
    
    # Extract regressors
    X_reg = pdf[x_cols].values
    Y = pdf["_mean_y"].values
    wts = pdf["_wts"].values
    
    # Build FE dummies using sparse matrix
    fe_dummies = []
    fe_col_names = []
    n_fe_levels = 0
    
    for fe in fe_cols:
        categories = pdf[fe].unique()
        n_cats = len(categories)
        n_fe_levels += n_cats
        
        # Create mapping
        cat_to_idx = {cat: i for i, cat in enumerate(categories)}
        
        # Create dummy matrix (drop first category for identification)
        for i, cat in enumerate(categories[1:], 1):
            col_name = f"{fe}_{cat}"
            dummy = (pdf[fe] == cat).astype(float).values
            fe_dummies.append(dummy)
            fe_col_names.append(col_name)
    
    # Combine regressors and FE dummies
    if fe_dummies:
        X_fe = np.column_stack(fe_dummies)
        X = np.hstack([X_reg, X_fe])
        all_col_names = list(x_cols) + fe_col_names
    else:
        X = X_reg
        all_col_names = list(x_cols)
    
    return X, Y, wts, all_col_names, n_fe_levels


def solve_wls(
    X: np.ndarray,
    Y: np.ndarray,
    wts: np.ndarray
) -> Tuple[np.ndarray, np.ndarray]:
    """
    Solve weighted least squares.
    
    Parameters
    ----------
    X : np.ndarray
        Design matrix (G x p)
    Y : np.ndarray
        Response vector (G,) - group means
    wts : np.ndarray
        Weights (G,) - sqrt(n_g)
        
    Returns
    -------
    tuple
        (beta, XtX_inv)
    """
    # Weight the design matrix and response
    Xw = X * wts[:, np.newaxis]
    Yw = Y * wts
    
    # Solve normal equations
    XtX = Xw.T @ Xw
    Xty = Xw.T @ Yw
    
    # Use Cholesky if possible, fallback to QR
    try:
        L = np.linalg.cholesky(XtX)
        beta = np.linalg.solve(L.T, np.linalg.solve(L, Xty))
        XtX_inv = np.linalg.solve(L.T, np.linalg.solve(L, np.eye(XtX.shape[0])))
    except np.linalg.LinAlgError:
        # Fallback to QR
        beta, _, _, _ = np.linalg.lstsq(XtX, Xty, rcond=None)
        XtX_inv = np.linalg.pinv(XtX)
    
    return beta, XtX_inv


def compute_rss_grouped(
    compressed_df,
    X: np.ndarray,
    beta: np.ndarray,
    backend: str = "polars"
) -> Tuple[float, np.ndarray]:
    """
    Compute RSS from sufficient statistics.
    
    RSS = sum_g (sum_y_sq_g - 2 * yhat_g * sum_y_g + n_g * yhat_g^2)
    
    Parameters
    ----------
    compressed_df : DataFrame
        Compressed data with _n, _sum_y, _sum_y_sq
    X : np.ndarray
        Design matrix
    beta : np.ndarray
        Coefficient vector
    backend : str
        "polars" or "duckdb"
        
    Returns
    -------
    tuple
        (rss_total, rss_g) - total RSS and per-group RSS
    """
    if backend == "polars":
        n_g = compressed_df["_n"].to_numpy()
        sum_y_g = compressed_df["_sum_y"].to_numpy()
        sum_y_sq_g = compressed_df["_sum_y_sq"].to_numpy()
    else:
        n_g = compressed_df["_n"].values
        sum_y_g = compressed_df["_sum_y"].values
        sum_y_sq_g = compressed_df["_sum_y_sq"].values
    
    # Fitted values for each group
    yhat_g = X @ beta
    
    # Per-group RSS
    rss_g = sum_y_sq_g - 2 * yhat_g * sum_y_g + n_g * (yhat_g ** 2)
    rss_total = np.sum(rss_g)
    
    return rss_total, rss_g


def compute_se_compress(
    XtX_inv: np.ndarray,
    rss_total: float,
    rss_g: np.ndarray,
    n_obs: int,
    df_resid: int,
    vcov: str,
    X: np.ndarray,
    x_cols: List[str]
) -> np.ndarray:
    """
    Compute standard errors from compressed data.
    
    Parameters
    ----------
    XtX_inv : np.ndarray
        Inverse of X'X
    rss_total : float
        Total residual sum of squares
    rss_g : np.ndarray
        Per-group RSS
    n_obs : int
        Original number of observations
    df_resid : int
        Residual degrees of freedom
    vcov : str
        "iid" or "HC1"
    X : np.ndarray
        Design matrix (compressed)
    x_cols : list of str
        Names of regressor columns (to extract subset of SEs)
        
    Returns
    -------
    np.ndarray
        Standard errors for x_cols only
    """
    k_x = len(x_cols)
    
    if vcov == "iid":
        sigma2 = rss_total / df_resid
        se_full = np.sqrt(np.diag(XtX_inv) * sigma2)
    elif vcov.upper() == "HC1":
        # Meat matrix: X' diag(rss_g) X
        meat = X.T @ (X * rss_g[:, np.newaxis])
        vcov_matrix = XtX_inv @ meat @ XtX_inv
        # HC1 adjustment
        adjustment = n_obs / df_resid
        se_full = np.sqrt(np.diag(vcov_matrix) * adjustment)
    else:
        raise ValueError(f"vcov must be 'iid' or 'HC1' for compress strategy, got '{vcov}'")
    
    # Return only SEs for x_cols (not FE dummies)
    return se_full[:k_x]


def leanfe_compress_polars(
    df: pl.DataFrame,
    y_col: str,
    x_cols: List[str],
    fe_cols: List[str],
    weights: Optional[str] = None,
    vcov: str = "iid"
) -> Dict:
    """
    Run compressed regression using Polars backend.
    
    Parameters
    ----------
    df : pl.DataFrame
        Input data
    y_col : str
        Dependent variable
    x_cols : list of str
        Regressors
    fe_cols : list of str
        Fixed effects
    weights : str, optional
        Weight column
    vcov : str
        "iid" or "HC1"
        
    Returns
    -------
    dict
        Regression results
    """
    # Compress data
    compressed, n_obs = compress_polars(df, y_col, x_cols, fe_cols, weights)
    n_compressed = len(compressed)
    
    # Build design matrix
    X, Y, wts, all_cols, n_fe_levels = build_design_matrix(
        compressed, x_cols, fe_cols, backend="polars"
    )
    
    # Solve WLS
    beta, XtX_inv = solve_wls(X, Y, wts)
    
    # Compute RSS
    rss_total, rss_g = compute_rss_grouped(compressed, X, beta, backend="polars")
    
    # Degrees of freedom
    p = len(all_cols)
    df_resid = n_obs - p
    
    # Standard errors
    se = compute_se_compress(XtX_inv, rss_total, rss_g, n_obs, df_resid, vcov, X, x_cols)
    
    # Extract coefficients for x_cols only
    beta_x = beta[:len(x_cols)]
    
    return {
        "coefficients": dict(zip(x_cols, beta_x)),
        "std_errors": dict(zip(x_cols, se)),
        "n_obs": n_obs,
        "n_compressed": n_compressed,
        "compression_ratio": n_compressed / n_obs,
        "vcov_type": vcov,
        "strategy": "compress",
        "df_resid": df_resid,
        "rss": rss_total,
    }


def leanfe_compress_duckdb(
    con,
    y_col: str,
    x_cols: List[str],
    fe_cols: List[str],
    weights: Optional[str] = None,
    vcov: str = "iid"
) -> Dict:
    """
    Run compressed regression using DuckDB backend.
    
    Parameters
    ----------
    con : duckdb.Connection
        DuckDB connection with 'data' table
    y_col : str
        Dependent variable
    x_cols : list of str
        Regressors
    fe_cols : list of str
        Fixed effects
    weights : str, optional
        Weight column
    vcov : str
        "iid" or "HC1"
        
    Returns
    -------
    dict
        Regression results
    """
    # Compress data
    compressed, n_obs = compress_duckdb(con, y_col, x_cols, fe_cols, weights)
    n_compressed = len(compressed)
    
    # Build design matrix
    X, Y, wts, all_cols, n_fe_levels = build_design_matrix(
        compressed, x_cols, fe_cols, backend="duckdb"
    )
    
    # Solve WLS
    beta, XtX_inv = solve_wls(X, Y, wts)
    
    # Compute RSS
    rss_total, rss_g = compute_rss_grouped(compressed, X, beta, backend="duckdb")
    
    # Degrees of freedom
    p = len(all_cols)
    df_resid = n_obs - p
    
    # Standard errors
    se = compute_se_compress(XtX_inv, rss_total, rss_g, n_obs, df_resid, vcov, X, x_cols)
    
    # Extract coefficients for x_cols only
    beta_x = beta[:len(x_cols)]
    
    return {
        "coefficients": dict(zip(x_cols, beta_x)),
        "std_errors": dict(zip(x_cols, se)),
        "n_obs": n_obs,
        "n_compressed": n_compressed,
        "compression_ratio": n_compressed / n_obs,
        "vcov_type": vcov,
        "strategy": "compress",
        "df_resid": df_resid,
        "rss": rss_total,
    }

File: r/R/compress.R
#' YOCO (You Only Compress Once) compression strategy for leanfe
#'
#' Implements the optimal data compression strategy from Wong et al. (2021):
#' "You Only Compress Once: Optimal Data Compression for Estimating Linear Models"
#'
#' Used automatically when vcov is "iid" or "HC1" (not cluster) and no IV.

#' @importFrom polars pl
#' @importFrom DBI dbGetQuery
NULL


#' Check if compression strategy should be used
#'
#' @param vcov Variance-covariance type
#' @param has_instruments Whether IV/2SLS is being used
#' @return Logical
#' @keywords internal
.should_use_compress <- function(vcov, has_instruments) {
  vcov_ok <- tolower(vcov) %in% c("iid", "hc1")
  return(vcov_ok && !has_instruments)
}


#' Compress data using Polars GROUP BY
#'
#' @param df Polars DataFrame
#' @param y_col Dependent variable
#' @param x_cols Regressors
#' @param fe_cols Fixed effects
#' @param weights Weight column (optional)
#' @return List with compressed df and n_obs_original
#' @keywords internal
.compress_polars <- function(df, y_col, x_cols, fe_cols, weights = NULL) {
  group_cols <- c(x_cols, fe_cols)
  n_obs_original <- df$height
  
  if (!is.null(weights)) {
    agg_exprs <- list(
      pl$col(weights)$sum()$alias("_n"),
      (pl$col(y_col) * pl$col(weights))$sum()$alias("_sum_y"),
      (pl$col(y_col)$pow(2) * pl$col(weights))$sum()$alias("_sum_y_sq")
    )
  } else {
    agg_exprs <- list(
      pl$len()$alias("_n"),
      pl$col(y_col)$sum()$alias("_sum_y"),
      pl$col(y_col)$pow(2)$sum()$alias("_sum_y_sq")
    )
  }
  
  compressed <- df$group_by(group_cols)$agg(!!!agg_exprs)
  
  # Add mean_y and sqrt weights for WLS
  compressed <- compressed$with_columns(
    (pl$col("_sum_y") / pl$col("_n"))$alias("_mean_y"),
    pl$col("_n")$sqrt()$alias("_wts")
  )
  
  list(compressed = compressed, n_obs_original = n_obs_original)
}


#' Compress data using DuckDB SQL
#'
#' @param con DuckDB connection
#' @param y_col Dependent variable
#' @param x_cols Regressors
#' @param fe_cols Fixed effects
#' @param weights Weight column (optional)
#' @return List with compressed df and n_obs_original
#' @keywords internal
.compress_duckdb <- function(con, y_col, x_cols, fe_cols, weights = NULL) {
  group_cols <- c(x_cols, fe_cols)
  group_cols_sql <- paste(group_cols, collapse = ", ")
  
  n_obs_original <- dbGetQuery(con, "SELECT COUNT(*) FROM data")[[1]]
  
  if (!is.null(weights)) {
    query <- sprintf("
      SELECT
        %s,
        SUM(%s) AS _n,
        SUM(%s * %s) AS _sum_y,
        SUM(POWER(%s, 2) * %s) AS _sum_y_sq,
        SUM(%s * %s) / SUM(%s) AS _mean_y,
        SQRT(SUM(%s)) AS _wts
      FROM data
      GROUP BY %s
    ", group_cols_sql, weights, y_col, weights, y_col, weights, 
       y_col, weights, weights, weights, group_cols_sql)
  } else {
    query <- sprintf("
      SELECT
        %s,
        COUNT(*) AS _n,
        SUM(%s) AS _sum_y,
        SUM(POWER(%s, 2)) AS _sum_y_sq,
        SUM(%s) / COUNT(*) AS _mean_y,
        SQRT(COUNT(*)) AS _wts
      FROM data
      GROUP BY %s
    ", group_cols_sql, y_col, y_col, y_col, group_cols_sql)
  }
  
  compressed_df <- dbGetQuery(con, query)
  list(compressed = compressed_df, n_obs_original = n_obs_original)
}


#' Build design matrix from compressed data
#'
#' @param compressed_df Compressed data frame
#' @param x_cols Regressors
#' @param fe_cols Fixed effects
#' @return List with X, Y, wts, all_col_names, n_fe_levels
#' @keywords internal
.build_design_matrix <- function(compressed_df, x_cols, fe_cols) {
  # Convert to data.frame if polars
  if ("polars_data_frame" %in% class(compressed_df)) {
    pdf <- as.data.frame(compressed_df)
  } else {
    pdf <- compressed_df
  }
  
  # Extract regressors
  X_reg <- as.matrix(pdf[, x_cols, drop = FALSE])
  Y <- pdf[["_mean_y"]]
  wts <- pdf[["_wts"]]
  
  # Build FE dummies
  fe_dummies <- list()
  fe_col_names <- c()
  n_fe_levels <- 0
  
  for (fe in fe_cols) {
    categories <- unique(pdf[[fe]])
    n_cats <- length(categories)
    n_fe_levels <- n_fe_levels + n_cats
    
    # Create dummies (drop first category for identification)
    for (i in seq_along(categories)[-1]) {
      cat <- categories[i]
      col_name <- paste0(fe, "_", cat)
      dummy <- as.numeric(pdf[[fe]] == cat)
      fe_dummies[[col_name]] <- dummy
      fe_col_names <- c(fe_col_names, col_name)
    }
  }
  
  # Combine regressors and FE dummies
  if (length(fe_dummies) > 0) {
    X_fe <- do.call(cbind, fe_dummies)
    X <- cbind(X_reg, X_fe)
    all_col_names <- c(x_cols, fe_col_names)
  } else {
    X <- X_reg
    all_col_names <- x_cols
  }
  
  list(X = X, Y = Y, wts = wts, all_col_names = all_col_names, n_fe_levels = n_fe_levels)
}


#' Solve weighted least squares
#'
#' @param X Design matrix
#' @param Y Response vector (group means)
#' @param wts Weights (sqrt(n_g))
#' @return List with beta and XtX_inv
#' @keywords internal
.solve_wls <- function(X, Y, wts) {
  # Weight the design matrix and response
  Xw <- X * wts
  Yw <- Y * wts
  
  # Solve normal equations
  XtX <- crossprod(Xw)
  Xty <- crossprod(Xw, Yw)
  
  # Use Cholesky if possible
  XtX_inv <- tryCatch({
    chol2inv(chol(XtX))
  }, error = function(e) {
    solve(XtX)
  })
  
  beta <- XtX_inv %*% Xty
  
  list(beta = as.vector(beta), XtX_inv = XtX_inv)
}


#' Compute RSS from sufficient statistics
#'
#' @param compressed_df Compressed data
#' @param X Design matrix
#' @param beta Coefficients
#' @return List with rss_total and rss_g
#' @keywords internal
.compute_rss_grouped <- function(compressed_df, X, beta) {
  # Convert to data.frame if polars
  if ("polars_data_frame" %in% class(compressed_df)) {
    pdf <- as.data.frame(compressed_df)
  } else {
    pdf <- compressed_df
  }
  
  n_g <- pdf[["_n"]]
  sum_y_g <- pdf[["_sum_y"]]
  sum_y_sq_g <- pdf[["_sum_y_sq"]]
  
  # Fitted values for each group
  yhat_g <- as.vector(X %*% beta)
  
  # Per-group RSS
  rss_g <- sum_y_sq_g - 2 * yhat_g * sum_y_g + n_g * (yhat_g^2)
  rss_total <- sum(rss_g)
  
  list(rss_total = rss_total, rss_g = rss_g)
}


#' Compute standard errors from compressed data
#'
#' @param XtX_inv Inverse of X'X
#' @param rss_total Total RSS
#' @param rss_g Per-group RSS
#' @param n_obs Original number of observations
#' @param df_resid Residual degrees of freedom
#' @param vcov "iid" or "HC1"
#' @param X Design matrix
#' @param k_x Number of regressor columns
#' @return Standard errors for x_cols only
#' @keywords internal
.compute_se_compress <- function(XtX_inv, rss_total, rss_g, n_obs, df_resid, vcov, X, k_x) {
  if (tolower(vcov) == "iid") {
    sigma2 <- rss_total / df_resid
    se_full <- sqrt(diag(XtX_inv) * sigma2)
  } else if (tolower(vcov) == "hc1") {
    # Meat matrix: X' diag(rss_g) X
    meat <- crossprod(X, X * rss_g)
    vcov_matrix <- XtX_inv %*% meat %*% XtX_inv
    # HC1 adjustment
    adjustment <- n_obs / df_resid
    se_full <- sqrt(diag(vcov_matrix) * adjustment)
  } else {
    stop(sprintf("vcov must be 'iid' or 'HC1' for compress strategy, got '%s'", vcov))
  }
  
  # Return only SEs for x_cols (not FE dummies)
  se_full[1:k_x]
}


#' Run compressed regression using Polars backend
#'
#' @param df Polars DataFrame
#' @param y_col Dependent variable
#' @param x_cols Regressors
#' @param fe_cols Fixed effects
#' @param weights Weight column (optional)
#' @param vcov "iid" or "HC1"
#' @return Regression results list
#' @keywords internal
.leanfe_compress_polars <- function(df, y_col, x_cols, fe_cols, weights = NULL, vcov = "iid") {
  # Compress data
  comp_result <- .compress_polars(df, y_col, x_cols, fe_cols, weights)
  compressed <- comp_result$compressed
  n_obs <- comp_result$n_obs_original
  n_compressed <- compressed$height
  
  # Build design matrix
  dm_result <- .build_design_matrix(compressed, x_cols, fe_cols)
  X <- dm_result$X
  Y <- dm_result$Y
  wts <- dm_result$wts
  all_cols <- dm_result$all_col_names
  
  # Solve WLS
  wls_result <- .solve_wls(X, Y, wts)
  beta <- wls_result$beta
  XtX_inv <- wls_result$XtX_inv
  
  # Compute RSS
  rss_result <- .compute_rss_grouped(compressed, X, beta)
  rss_total <- rss_result$rss_total
  rss_g <- rss_result$rss_g
  
  # Degrees of freedom
  p <- length(all_cols)
  df_resid <- n_obs - p
  
  # Standard errors
  se <- .compute_se_compress(XtX_inv, rss_total, rss_g, n_obs, df_resid, vcov, X, length(x_cols))
  
  # Extract coefficients for x_cols only
  beta_x <- beta[1:length(x_cols)]
  
  list(
    coefficients = setNames(beta_x, x_cols),
    std_errors = setNames(se, x_cols),
    n_obs = n_obs,
    n_compressed = n_compressed,
    compression_ratio = n_compressed / n_obs,
    vcov_type = vcov,
    strategy = "compress",
    df_resid = df_resid,
    rss = rss_total
  )
}


#' Run compressed regression using DuckDB backend
#'
#' @param con DuckDB connection
#' @param y_col Dependent variable
#' @param x_cols Regressors
#' @param fe_cols Fixed effects
#' @param weights Weight column (optional)
#' @param vcov "iid" or "HC1"
#' @return Regression results list
#' @keywords internal
.leanfe_compress_duckdb <- function(con, y_col, x_cols, fe_cols, weights = NULL, vcov = "iid") {
  # Compress data
  comp_result <- .compress_duckdb(con, y_col, x_cols, fe_cols, weights)
  compressed <- comp_result$compressed
  n_obs <- comp_result$n_obs_original
  n_compressed <- nrow(compressed)
  
  # Build design matrix
  dm_result <- .build_design_matrix(compressed, x_cols, fe_cols)
  X <- dm_result$X
  Y <- dm_result$Y
  wts <- dm_result$wts
  all_cols <- dm_result$all_col_names
  
  # Solve WLS
  wls_result <- .solve_wls(X, Y, wts)
  beta <- wls_result$beta
  XtX_inv <- wls_result$XtX_inv
  
  # Compute RSS
  rss_result <- .compute_rss_grouped(compressed, X, beta)
  rss_total <- rss_result$rss_total
  rss_g <- rss_result$rss_g
  
  # Degrees of freedom
  p <- length(all_cols)
  df_resid <- n_obs - p
  
  # Standard errors
  se <- .compute_se_compress(XtX_inv, rss_total, rss_g, n_obs, df_resid, vcov, X, length(x_cols))
  
  # Extract coefficients for x_cols only
  beta_x <- beta[1:length(x_cols)]
  
  list(
    coefficients = setNames(beta_x, x_cols),
    std_errors = setNames(se, x_cols),
    n_obs = n_obs,
    n_compressed = n_compressed,
    compression_ratio = n_compressed / n_obs,
    vcov_type = vcov,
    strategy = "compress",
    df_resid = df_resid,
    rss = rss_total
  )
}
